### ISSUE TITLE
Add --update-signature option (was Possibility to update files only if filename changed?)

### ISSUE THREAD

**Issue by narensankar0529** [RhetTbull/osxphotos#286]
Hi

I am thinking of two scenarios and wondering if it is possible to support this - 

1. I would like to convert all of my old h.264 encoded videos in Photos to h.265 when they are exported (helps in saving cloud storage space for backup). If I do the outside of the osxphotos workflow, then when I run --update again, osxphotos overwrites the re-encoded videos since they are different on disk. Is there a way to tell osxphotos to not do this?

2. I am trying out Mylio as a distributed sync tool and they have a problem to handle the dates on videos correctly. They look at modification date v/s createdDate. So again if I use exiftool to fix the modification date externally osxphotos will again overwrite this on the next update.

So is there a way to tell osxphotos to only update a file if the file a) does not exist b) was deleted in the photos library c) perhaps have a size of 0 and in all other cases as long as there is a file with the same name as the template just skip and move on to the next and ignore the file modifications on the output disk?

**Comment by narensankar0529** [comment]
Basically allow to run an export ignoring your stored size and modification date information in the DB. Perhaps a flag that says --filename-only

**Comment by RhetTbull** [comment]
There's currently no way of running an update that only looks filename though it could be added.  This would make the already complex export logic more complicated though.  

For scenario #2, I'm currently working on another issue #282) regarding --exiftool and dates on videos so once I resolve this, it might not be an issue.  What command would you run with exiftool to fix the file modification date?

For scenario #1, how are you converting the videos?  Is it a command line app?  I've thought about adding a feature to run an external command then update the file signature after it's complete so that --update would work as expected.

How comfortable are you with python?  You could write a custom export script yourself using the osxphotos API.  There's a couple of [examples](https://github.com/RhetTbull/osxphotos/blob/master/examples/) in the repo.

**Comment by RhetTbull** [comment]
Also, I've got another open issue #243 I'm working on to add a new `--exiftool-template` option that would allow you to run arbitrary exiftool commands on the exported photo and have the resulting signature stored in the database so --update works.  This would address your second scenario if the fix I'm working on for #282 doesn't address it.

**Comment by narensankar0529** [comment]
For scenario #2 I am using exiftool - exiftool "-FileModifyDate<MediaCreateDate" filename

For scenario #1 I have a bash script that uses ffprobe to check if the codec is not HEVC and then uses FFmpeg to re-encode to HEVC so having the ability to run an arbitrary script or command would be great. Unfortunately I am only handy with C or bash scripting and never really picked up python

**Comment by RhetTbull** [comment]
For scenario #2 wouldn't --touch-file do that for you?

For scenario #1, I was thinking about this more and there are a couple of options.  1) do as I first suggested where there's an option to run an arbitrary command then look at the file after the command is run to update the signature database (this would assume the command you run leaves the filename the same) or 2) add a --post-process flag that tells osxphotos you did post-processing and to update all the signatures so that next time you run --update, it looks at the new signatures.

**Comment by narensankar0529** [comment]
Scenario #2 - yes this would work. I did not notice that option. Would that work on videos as well?

Actually perhaps --update-signatures would be a better option since then it is easier to do bulk transformations.

Right now what I do is -

Step 1 - download missing
Step 2 - export media
Step 3 - fix up permissions on media for other apps to be able to sync to media

So if I can do a step 2a which does transcoding, and then step 2b that does update-signatures, it would be more efficient I think. Can run multiple transcodes in parallel v/s having to do it one file at a time when exporting.

**Comment by RhetTbull** [comment]
Yes, --touch-file works on videos. It sets the file modification time to match the creation date of the asset. It also works with --update. The other issue I'm working on is related to the --exittool option which isn't always updating the QuickTime time tags for videos.

I'll work on adding a --update-signature option when I finish the current issues I'm working on. You could then do your post-processing then do something like `osxphotos export /path --update-signatures` then running --update would work.

**Comment by narensankar0529** [comment]
Awesome! Thanks a lot

**Comment by narensankar0529** [comment]
One more question - I assume --touch-file will only work if I don't use --export-as-hardlink - correct?

**Comment by RhetTbull** [comment]
No, --touch-file works fine with --export-as-hardlink.  Photos doesn't complain if the file mod time is changed in the Photos library.

**Comment by narensankar0529** [comment]
Great to know this. So makes the export simpler to implement.

Thanks!
—————

Naren Sankar

**Comment by RhetTbull** [comment]
After looking at the code, I realize that for original photos, it doesn't store the signature -- it compares the original in the Photos library and the exported version.  All others (edited, exiftool, converted) store a signature because some change has occurred that osxphotos won't have a priori knowledge of for future comparisons.  Thus my original proposal to have an `--update-signature` option won't work in the case of original photos.  Instead I've implemented a `--ignore-signatures` option which does as you originally proposed and only looks at filenames, not signatures, for all exported files.  I need to add tests as there are several export options that need to be tested in different combinations with this (e.g. `--touch-file`)

The advantage of a `--update-signatures` option would be that you could do your post-processing, run `--update-signatures` and effectively "freeze" the export database so any other changes (eg. unintentional edits) would be caught by osxphotos on next update.  However, doing this would require substantial changes to the export code.  Implementing `--ignore-signatures` on the other hand took only a few lines of code.

**Comment by RhetTbull** [comment]
--ignore-signature  implemented in release [0.38.5](https://github.com/RhetTbull/osxphotos/releases/tag/v0.38.5) which should provide a solution for scenario #1.

**Comment by narensankar0529** [comment]
Thanks for the implementation. One question - sidecar files are not updated if the image file is not updated - correct? But is sidecar does not exist will the sidecar be written or ignored?

**Comment by RhetTbull** [comment]
Processing of sidecar files is independent of the image files.  The export routine looks to see if they metadata going into the sidecar (dates, location, persons, keywords, etc) has changed (basically, the code generates the sidecar then produces a hash of it and compares that hash with the sidecar hash stored in the export database), and if so writes the sidecar otherwise skips it. I've not tested it but the expected behavior between sidecars and --ignore-signature is this:

Case: export with --sidecar and then you update the exported photo files then run export again with --ignore-signature and --sidecar
- If the metadata (in Photos) that went into the sidecar did not change, the sidecar will not be updated
- If the metadata (in Photos) that went into the sidecar did change, a new sidecar is written but a new image file is not
- If a sidecar does not exist for the photo, a sidecar will be written whether or not the photo file was written

**Comment by RhetTbull** [comment]
I'll add a test case to confirm this.  If you notice differently, let me know.

**Comment by narensankar0529** [comment]
Great! This look ok so far.

Thanks!
—————

Naren Sankar

**Comment by RhetTbull** [comment]
I've confirmed the behavior I described above is what happens and I added test cases for these.

**Comment by narensankar0529** [comment]
One suggestion - Could you add this as documentation in the help section of osxphotos help export? This will explicitly document how this all works -

Case: export with --sidecar and then you update the exported photo files then run export again with --ignore-signature and --sidecar

If the metadata (in Photos) that went into the sidecar did not change, the sidecar will not be updated
If the metadata (in Photos) that went into the sidecar did change, a new sidecar is written but a new image file is not
If a sidecar does not exist for the photo, a sidecar will be written whether or not the photo file was written

**Comment by RhetTbull** [comment]
Done! I won't push a new release until the next feature release but the docs are updated in the "master" branch on github.
